tests/py_localhost/server.py

```diff
@@ -0,0 +1,358 @@
+
+)
+
+from h2.errors import ErrorCodes
+
+from h2.exceptions import ProtocolError, StreamClosedError
+
+from h2.settings import SettingCodes
+
+
+RequestData = collections.namedtuple('RequestData', ['headers', 'data'])
+
+
+class H2Protocol(asyncio.Protocol):
+
+    def __init__(self):
+
+        config = H2Configuration(client_side=False, header_encoding='utf-8')
+
+        self.conn = H2Connection(config=config)
+
+        self.transport = None
+
+        self.stream_data = {}
+
+        self.flow_control_futures = {}
+
+    def connection_made(self, transport: asyncio.Transport):
+
+        self.transport = transport
+
+        self.conn.initiate_connection()
+
+        self.transport.write(self.conn.data_to_send())
+
+    def connection_lost(self, exc):
+
+        for future in self.flow_control_futures.values():
+
+            future.cancel()
+
+        self.flow_control_futures = {}
+
+    def data_received(self, data: bytes):
+
+        try:
+
+            events = self.conn.receive_data(data)
+
+        except ProtocolError as e:
+
+            self.transport.write(self.conn.data_to_send())
+
+            self.transport.close()
+
+        else:
+
+            self.transport.write(self.conn.data_to_send())
+
+            for event in events:
+
+                if isinstance(event, RequestReceived):
+
+                    self.request_received(event.headers, event.stream_id)
+
+                elif isinstance(event, DataReceived):
+
+                    self.receive_data(event.data, event.stream_id)
+
+                elif isinstance(event, StreamEnded):
+
+                    self.stream_complete(event.stream_id)
+
+                elif isinstance(event, ConnectionTerminated):
+
+                    self.transport.close()
+
+                elif isinstance(event, StreamReset):
+
+                    self.stream_reset(event.stream_id)
+
+                elif isinstance(event, WindowUpdated):
+
+                    self.window_updated(event.stream_id, event.delta)
+
+                elif isinstance(event, RemoteSettingsChanged):
+
+                    if SettingCodes.INITIAL_WINDOW_SIZE in event.changed_settings:
+
+                        self.window_updated(None, 0)
+
+                self.transport.write(self.conn.data_to_send())
+
+    def request_received(self, headers: List[Tuple[str, str]], stream_id: int):
+        self.raw_headers = headers
+
+        headers = collections.OrderedDict(headers)
+
+        # Store off the request data.
+
+        request_data = RequestData(headers, io.BytesIO())
+
+        self.stream_data[stream_id] = request_data
+
+    def stream_complete(self, stream_id: int):
+        """
+
+        When a stream is complete, we can send our response.
+
+        """
+
+        try:
+
+            request_data = self.stream_data[stream_id]
+
+        except KeyError:
+
+            # Just return, we probably 405'd this already
+
+            return
+
+        response_headers = [(':status', '200')]
```

Comment:
right now the "echo" behavior hard-coded as the only possible response.
can you edit this to be a little more flexible, like check the `:path` and redirect to a specific function, so we can easily add more tests in the future?

something like:
```
...
path = request_data.headers[':path'] 
if path == '/echo':
    self.handle_request_echo(request_data)
else:
    send 404 response
...


```